lib/rmt/file_validator.rb

```diff
@@ -0,0 +1,59 @@
+require 'fileutils'
+require 'rmt/checksum_verifier'
+
+class RMT::FileValidator
+  class << self
+    def validate_local_file(file_reference, deep_verify:)
+      verify_file_on_disk(file_reference, deep_verify).tap do |valid_on_disk|
+        update_file_status_on_database(file_reference, valid_on_disk)
+      end
+    end
+
+    def find_valid_file_by_checksum(checksum, checksum_type, deep_verify:)
+      ::DownloadedFile
+        .where(checksum: checksum, checksum_type: checksum_type)
+        .map { |file| [file, validate_tracked_file_on_disk(file, deep_verify)] }
+        .find { |_, valid_on_disk| valid_on_disk }
```

Comment:
Instead of mapping an array of file records into a 2-dimensional array -- I'd consider expanding `.map` and `.find` into a regular `.each` loop.

This mega-chain of 5 methods isn't making things any easier to understand.

Reply:
In fact, I don't think `.each` makes things less complicated: `.each` would have the same function as `.map`, to apply `validate_tracked_file_on_disk` on **each** file. In the end, I'd still need to get rid of all invalid files on disk and choose one valid file to returns its path. I agree the two-dimensional array can be a bit confusing, though Ruby hashes are iterated the same way.

I can write something like this to remove one level of chaining:

```ruby
      ::DownloadedFile
        .where(checksum: checksum, checksum_type: checksum_type)
        .select { |file| validate_tracked_file_on_disk(file, deep_verify) }
        &.first&.local_path
```

However, I'm not figuring out how to use `.each` to make it more readable. Do you have any suggestions?

Reply:
I can open a PR with proposed changes, but that's going to take me a while.

Reply:
@lcaparroz The dual responsibility of deleting invalid files while also finding a valid instance makes the method confusing. 

If we need to keep it that way, here a potential `.each` block example.

```ruby
def find_valid_file_by_checksum(checksum, checksum_type, deep_verify:)
  file_path = nil
  ::DownloadedFile.where(checksum: checksum, checksum_type: checksum_type).each do |file|
    file_path ||= file.local_path if validate_tracked_file_on_disk(file, deep_verify)
  end
  local
end
```

@ikapelyukhin Is something like this what you're looking for?

Reply:
Check the last commit, I rewrote the method and maybe it got better to understand it.

Reply:
I guess we have some consensus now, the final implementation for this method is [here](https://github.com/SUSE/rmt/pull/594/commits/c417d92637b04ebba6f88faddb985e90fdb7df60#diff-626359a9598a29f883f7227ee01a261bR21-R33).